src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,97 @@
+        }
+
+        public Task DownloadMultipleArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            return this.DownloadFileShareAsync(downloadParameters.ProjectId, buildArtifacts, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public Task DownloadFileShareAsync(Guid projectId, IEnumerable<BuildArtifact> buildArtifacts, string targetDirectory, IEnumerable<string> minimatchFilters, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                return DownloadFileShareAsync(projectId, buildArtifact, dirPath, minimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+
+            return Task.FromResult<object>(null);
+        }
+
+        private Task DownloadFileShareAsync(Guid projectId, BuildArtifact artifact, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken, bool isSingleArtifactDownload = true)
+        {
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            var downloadRootPath = artifact.Resource.Data + Path.DirectorySeparatorChar + artifact.Name;
+            return DirectoryCopyWithMiniMatch(downloadRootPath, destPath, this.context, 1, minimatcherFuncs, downloadRootPath);
+        }
+
+        internal static Task DirectoryCopyWithMiniMatch(string sourcePath, string destPath, AgentTaskPluginExecutionContext context, int parallelCount = 1, IEnumerable<Func<string, bool>> minimatchFuncs = null, string minimatchRoot = null)
+        {
+            // If the source path is a file, the system should copy the file to the dest directory directly. 
+            if(File.Exists(sourcePath)) {
+                context.Output(StringUtil.Loc("CopyFileToDestination", sourcePath, destPath));
+                File.Copy(sourcePath, destPath + Path.DirectorySeparatorChar + Path.GetFileName(sourcePath), true);
+                return Task.FromResult<object>(null);
+            }
+
+            var opts = new ParallelOptions() { MaxDegreeOfParallelism = parallelCount };
+
+            // Get the subdirectories for the specified directory.
+            DirectoryInfo dir = new DirectoryInfo(sourcePath);
+
+            // If the destination directory doesn't exist, create it.
+            if (!Directory.Exists(destPath))
+            {
+                Directory.CreateDirectory(destPath);
+            }
+
+            DirectoryInfo[] dirs = dir.GetDirectories();
+                
```

Comment:
I'm not getting this we're only copying the first level directories? I'm pretty sure you need to pass SearchOption.AllDirectories:
[https://docs.microsoft.com/en-us/dotnet/api/system.io.searchoption?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.io.searchoption?view=netframework-4.8)